### ISSUE TITLE
Runtime requirement

### ISSUE THREAD

**Issue by Demirrr** [ttrouill/complex#13]
Hello Théo,

firstly, I would like to genuinely thank you for the code and rep which are well prepared and straightforward. However, the following scenario confuses me:
Given the following settings,
nb entities: 521,358, relations:70 and obs triples:2,395,882 )
Learning rate: 0.5
Max iter: 100
Generated negatives ratio: 10
Batch size: 23958
Starting grid search on: Complex_Logistic_Model
train: 100 mini-batches from callable
downhill: compiling evaluation function
downhill: compiling ADAGRAD optimizer
downhill: setting: rms_regularizer = 1e-08
downhill: setting: patience = 9999999
downhill: setting: validate_every = 9999999
downhill: setting: min_improvement = 0
downhill: setting: max_gradient_norm = 1
downhill: setting: max_gradient_elem = 0
downhill: setting: learning_rate = TensorConstant{0.5}
downhill: setting: momentum = 0
downhill: setting: nesterov = False

, it takes hours to train models. This issue stems from the implementation or am I lacking of a theoretical background here?

Cheers

**Comment by ttrouill** [comment]
Hello!

That's pretty much what I would expect, with 10 negatives per positive, you have here about 25M samples per iteration, and you have a lot of entities which means a big entity matrix that will cause a lot of cache misses at execution due to the random nature of the sampling. This implementation uses theano, it means that the C or CUDA code (depending on whether you're running on CPU or GPU) corresponding to the update steps of the gradient descent is generated and compiled on the fly when you run the script. Moreover, when possible, theano calls BLAS routines for classic linear algebra operations, which, if possible and if your BLAS library supports it, are parallelized (however there are no such operations in the ComplEx model). 

Basically, it is quite likely that you are pretty close to the best you can get from one core of your CPU, or from your GPU. However, what this code doesn't allow for is a distributed SGD, which means computing updates of different samples in parallel on mutliple CPUs or GPUs. This is doable with tensorflow, so if you really need better performances and have many GPUs (because it's probably not really worth doing it on CPUs given the speed-up you already get with a single GPU), you can try reimplementing the complEx model in tensorflow (or maybe it already exists somewhere), and use its [distributed training features](https://www.tensorflow.org/guide/distribute_strategy).

Good luck!
